### PR TITLE
#1535 - Add Other Reasons to Reimbursement Products

### LINKED ISSUE #1535: Finance Fixes - Add "other" categories
Epic: #1274 

### Description

We want to add an option to add other categories in instead of projects on the product section of the reimbursement request form. dropdown should have a project/category option. When you select one it should open the product view with a description and cost, and then on the left should be a dropdown of options (either projects or categories depending on what you selected)

### PR DESCRIPTION
**Pull request by RChandler234**
## Changes

Added Other Reasons as Alternatives to Projects in the Reimbursement Product Section. This is for the case when an expense doesn't fall cleanly into a project/ is mostly to enable us to be able to track spending in these areas

## Todo

- [ ] probably fix some tests
- [ ] get frontend form functioning how I want

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #1535  (issue #)

### CODE REVIEW

File: src/backend/src/services/reimbursement-requests.services.ts [hunk 1]
```diff
@@ -150,21 +157,30 @@ export default class ReimbursementRequestService {
...
 
+    const reimbursementProductsPromises = validatedReimbursementProducts.map(async (product) => {
+      return await prisma.reimbursement_Product.create({
+        data: {
+          name: product.name,
+          cost: product.cost,
+          reimbursementRequestId: createdReimbursementRequest.reimbursementRequestId,
+          reimbursementProductReason: {
+            create: {
+              wbsElementId: (product.reason as { wbsNum: WbsNumber; wbsElementId: number }).wbsElementId,
```
Comment (Peyton-McKee) [inline src/backend/src/services/reimbursement-requests.services.ts:172]: type these

File: src/backend/src/utils/reimbursement-requests.utils.ts [hunk 2]
```diff
@@ -59,35 +65,44 @@ export const validateReimbursementProducts = async (
...
-    const wbsElement = await prisma.wBS_Element.findFirst({
-      where: {
-        carNumber: wbsNum.carNumber,
-        projectNumber: wbsNum.projectNumber,
-        workPackageNumber: wbsNum.workPackageNumber
-      }
-    });
-    if (!wbsElement) throw new NotFoundException('WBS Element', wbsPipe(wbsNum));
-    if (wbsElement.dateDeleted) throw new DeletedException('WBS Element', wbsPipe(wbsNum));
+    reason: { wbsElementId: number; wbsNum: WbsNumber } | OtherProductReason;
```
Comment (Peyton-McKee) [inline src/backend/src/utils/reimbursement-requests.utils.ts:77]: type this

Comment (Peyton-McKee) [reply src/backend/src/utils/reimbursement-requests.utils.ts:77]: We should make { wbs.... : wbs... } | OtherProductReason a type and just return that instead of oring it everywhere

File: src/backend/src/utils/reimbursement-requests.utils.ts [hunk 3]
```diff
@@ -59,35 +65,44 @@ export const validateReimbursementProducts = async (
...
+  }>[] = reasons.map(async (reason: ReimbursementProductReasonCreateArgs, index) => {
+    if (!!(reason as WbsNumber).carNumber) {
```
Comment (Peyton-McKee) [inline src/backend/src/utils/reimbursement-requests.utils.ts:79]: why .carNumber

File: src/backend/src/utils/reimbursement-requests.utils.ts [hunk 4]
```diff
@@ -190,15 +205,26 @@ const updateDeletedProducts = async (products: Reimbursement_Product[]) => {
...
-    const validatedReimbursementProudcts = await validateReimbursementProducts(products);
-    await prisma.reimbursement_Product.createMany({
-      data: validatedReimbursementProudcts.map((product) => ({
-        name: product.name,
-        cost: product.cost,
-        wbsElementId: product.wbsElementId,
-        reimbursementRequestId
-      }))
+    const validatedReimbursementProducts = await validateReimbursementProducts(products);
+    const reimbursementProductsPromises = validatedReimbursementProducts.map(async (product) => {
```
Comment (Peyton-McKee) [inline src/backend/src/utils/reimbursement-requests.utils.ts:209]: isnt this exact function used somewhere else?

File: src/backend/src/transformers/reimbursement-requests.transformer.ts [hunk 5]
```diff
@@ -73,11 +78,20 @@ export const reimbursementProductTransformer = (
     name: reimbursementProduct.name,
     dateDeleted: reimbursementProduct.dateDeleted ?? undefined,
     cost: reimbursementProduct.cost,
-    wbsNum: wbsNumOf(reimbursementProduct.wbsElement),
-    wbsName: reimbursementProduct.wbsElement.name
+    reimbursementProductReason: reimbursementProduct.reimbursementProductReason
+      ? reimbursementProductReasonTransformer(reimbursementProduct.reimbursementProductReason)
+      : OtherProductReason.UNKNOWN
```
Comment (Peyton-McKee) [inline src/backend/src/transformers/reimbursement-requests.transformer.ts:83]: why is this here?

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
I feel like im doing alot of reading the same code so just like making sure were abstracting away all this correctly to concentrate the ugly type casting and such to like one file or something

File: src/backend/src/utils/reimbursement-requests.utils.ts (hunk 3 above)
Comment (RChandler234) [reply src/backend/src/utils/reimbursement-requests.utils.ts:79]: basically just checking whether it's a wbsNum or an other reason, this is one of the things that's gonna be abstracted away to a method

**Review by callmejim1226 — CHANGES_REQUESTED** [review]
Overall, looks good to me. Great work! I checked all the code and reviewed the feature broadly on frontend, backend, and shared type definitions. 

I confirmed the feature works well in my local app, where I was able to create RR with multiple products in either wbsElement or one of the categories.

One thing I caught was we also want to update the label on RR products on the RR details page. From the screenshot, we should change the 'WBS Element' label to the 'Project/Category' we have on the RR edit page.

image

**Comment by RChandler234** [comment]
use enumPipe to fix how enums are being displayed
